### ISSUE TITLE
NGINX uses old IP after websocket restart

### ISSUE THREAD

**Issue by martinvonwittich** [zammad/zammad-docker-compose#96]
### Infos:

* Docker version: docker-ce 18.03.0~ce-0~debian
* Operating system (Docker host): Debian 9

### Expected behavior:

* WebSocket service shouldn't break when it (and only it) is restarted

### Actual behavior:

* The WebSocket container may randomly receive a new IP from Docker, but NGINX will continue to use the old one. Therefore WebSocket breaks completely.

Output of `docker logs --tail 10 -f zammaddockercompose_zammad-nginx_1`:

```
[...]
172.17.0.1 - - [09/Feb/2019:23:21:41 +0000] "GET /api/v1/tickets/41461?all=true&_=1549754498312 HTTP/1.1" 200 8393 "https://zammad.ourcompany.tld/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/71.0.3578.98 Chrome/71.0.3578.98 Safari/537.36"
2019/02/09 23:21:42 [error] 17#17: *33 connect() failed (113: No route to host) while connecting to upstream, client: 172.17.0.1, server: _, request: "GET /ws HTTP/1.1", upstream: "http://172.17.0.7:6042/ws", host: "localhost:8080"
172.17.0.1 - - [09/Feb/2019:23:21:42 +0000] "GET /ws HTTP/1.1" 502 575 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/71.0.3578.98 Chrome/71.0.3578.98 Safari/537.36"
[...]
```

As you can see, it's trying to connect to `172.17.0.7`. This was the old IP of the WebSocket container; it changed to 172.17.0.5 after the restart:

```
zammad.ourcompany.tld ~/zammad-docker-compose # docker inspect -f '{{.Name}} {{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' $(docker ps -aq)
/zammaddockercompose_zammad-nginx_1 172.17.0.10
/zammaddockercompose_zammad-scheduler_1 172.17.0.9
/zammaddockercompose_zammad-backup_1 172.17.0.8
/zammaddockercompose_zammad-websocket_1 172.17.0.7
/zammaddockercompose_zammad-railsserver_1 172.17.0.6
/zammaddockercompose_zammad-init_1
/zammaddockercompose_zammad-elasticsearch_1 172.17.0.4
/zammaddockercompose_zammad-memcached_1 172.17.0.3
/zammaddockercompose_zammad-postgresql_1 172.17.0.2

zammad.ourcompany.tld ~/zammad-docker-compose # docker-compose restart zammad-websocket                                               
Restarting zammaddockercompose_zammad-websocket_1 ... done

zammad.ourcompany.tld ~/zammad-docker-compose # docker inspect -f '{{.Name}} {{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' $(docker ps -aq)
/zammaddockercompose_zammad-nginx_1 172.17.0.10
/zammaddockercompose_zammad-scheduler_1 172.17.0.9
/zammaddockercompose_zammad-backup_1 172.17.0.8
/zammaddockercompose_zammad-websocket_1 172.17.0.5
/zammaddockercompose_zammad-railsserver_1 172.17.0.6
/zammaddockercompose_zammad-init_1
/zammaddockercompose_zammad-elasticsearch_1 172.17.0.4
/zammaddockercompose_zammad-memcached_1 172.17.0.3
/zammaddockercompose_zammad-postgresql_1 172.17.0.2
```

Apparently NGINX resolved the host names in its configuration only once. `ping` correctly resolves the host to the new IP:

```
zammad.ourcompany.tld ~ # docker exec -it zammaddockercompose_zammad-nginx_1 /bin/bash
root@6267832da383:/opt/zammad# ping -c1 zammad-websocket                                                                        
PING zammad-websocket (172.17.0.5) 56(84) bytes of data.
64 bytes from zammaddockercompose_zammad-websocket_1.zammaddockercompose_default (172.17.0.5): icmp_seq=1 ttl=64 time=0.108 ms

--- zammad-websocket ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.108/0.108/0.108/0.000 ms
root@6267832da383:/opt/zammad# exit
```

### Steps to reproduce the behavior:

* Restart the WebSocket container:

	```
	docker-compose restart zammad-websocket
	```
	
	Note that this works only when there is a free IP available that is:
	- greater than 172.17.0.1 (reserved by Docker itself) and
	- _less_ than the old IP
	
	As you can see in the output above, 172.17.0.5 was available before the restart, and so `zammaddockercompose_zammad-websocket_1` received this IP. Further restarts will _not_ change the IP address, because there is no lower IP available.
	
### Possible solution

This looks like a possible solution to get NGINX to re-resolve the hostnames: https://serverfault.com/a/593003/124717

**Comment by martinvonwittich** [comment]
In reply to https://github.com/zammad/zammad-docker/issues/29#issuecomment-462403160 : 

> Never seen this before.
> How long do you have the container running?

I'm not sure how long the containers where running when I noticed this for the first time. I've since restarted them several times to isolate this issue. I can reproduce it after the whole application has been running for about a minute.

I believe the root cause, next to the behavior of NGINX, is the `zammaddockercompose_zammad-init_1` container - it has an IP when the whole application starts, but loses it when the init container exits. Docker will then randomly assign this IP to the next container that restarts:

1. Completely stop the application:

	```
	zammad.ourcompany.tld ~/zammad-docker-compose # docker-compose down 
	Stopping zammaddockercompose_zammad-nginx_1 ... done
	Stopping zammaddockercompose_zammad-scheduler_1 ... done
	Stopping zammaddockercompose_zammad-backup_1 ... done
	Stopping zammaddockercompose_zammad-websocket_1 ... done
	Stopping zammaddockercompose_zammad-railsserver_1 ... done
	Stopping zammaddockercompose_zammad-elasticsearch_1 ... done
	Stopping zammaddockercompose_zammad-postgresql_1 ... done
	Stopping zammaddockercompose_zammad-memcached_1 ... done
	Removing zammaddockercompose_zammad-nginx_1 ... done
	Removing zammaddockercompose_zammad-scheduler_1 ... done
	Removing zammaddockercompose_zammad-backup_1 ... done
	Removing zammaddockercompose_zammad-websocket_1 ... done
	Removing zammaddockercompose_zammad-railsserver_1 ... done
	Removing zammaddockercompose_zammad-init_1 ... done
	Removing zammaddockercompose_zammad-elasticsearch_1 ... done
	Removing zammaddockercompose_zammad-postgresql_1 ... done
	Removing zammaddockercompose_zammad-memcached_1 ... done
	Removing network zammaddockercompose_default
	```

2. Restart it:

	```
	zammad.ourcompany.tld ~/zammad-docker-compose # docker-compose up -d
	Creating network "zammaddockercompose_default" with the default driver
	Creating zammaddockercompose_zammad-memcached_1
	Creating zammaddockercompose_zammad-postgresql_1
	Creating zammaddockercompose_zammad-elasticsearch_1
	Creating zammaddockercompose_zammad-init_1
	Creating zammaddockercompose_zammad-railsserver_1
	Creating zammaddockercompose_zammad-backup_1
	Creating zammaddockercompose_zammad-websocket_1
	Creating zammaddockercompose_zammad-scheduler_1
	Creating zammaddockercompose_zammad-nginx_1
	```

3. Monitor the IPs of the containers. I have a shell script for this purpose to make it easier:

	```
	zammad.ourcompany.tld ~ # cat =dip
	#!/bin/sh
	#docker inspect -f '{{.Name}} {{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' "$@"
	docker inspect -f '{{.Name}} {{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' $(docker ps -aq)
	```

	```
	zammad.ourcompany.tld ~ # dip
	/zammaddockercompose_zammad-nginx_1 172.17.0.10
	/zammaddockercompose_zammad-scheduler_1 172.17.0.9
	/zammaddockercompose_zammad-websocket_1 172.17.0.8
	/zammaddockercompose_zammad-backup_1 172.17.0.7
	/zammaddockercompose_zammad-railsserver_1 172.17.0.6
	/zammaddockercompose_zammad-init_1 172.17.0.5
	/zammaddockercompose_zammad-elasticsearch_1 172.17.0.4
	/zammaddockercompose_zammad-postgresql_1 172.17.0.3
	/zammaddockercompose_zammad-memcached_1 172.17.0.2
	zammad.ourcompany.tld ~ # dip
	/zammaddockercompose_zammad-nginx_1 172.17.0.10
	/zammaddockercompose_zammad-scheduler_1 172.17.0.9
	/zammaddockercompose_zammad-websocket_1 172.17.0.8
	/zammaddockercompose_zammad-backup_1 172.17.0.7
	/zammaddockercompose_zammad-railsserver_1 172.17.0.6
	/zammaddockercompose_zammad-init_1 172.17.0.5
	/zammaddockercompose_zammad-elasticsearch_1 172.17.0.4
	/zammaddockercompose_zammad-postgresql_1 172.17.0.3
	/zammaddockercompose_zammad-memcached_1 172.17.0.2
	zammad.ourcompany.tld ~ # dip
	/zammaddockercompose_zammad-nginx_1 172.17.0.10
	/zammaddockercompose_zammad-scheduler_1 172.17.0.9
	/zammaddockercompose_zammad-websocket_1 172.17.0.8
	/zammaddockercompose_zammad-backup_1 172.17.0.7
	/zammaddockercompose_zammad-railsserver_1 172.17.0.6
	/zammaddockercompose_zammad-init_1
	/zammaddockercompose_zammad-elasticsearch_1 172.17.0.4
	/zammaddockercompose_zammad-postgresql_1 172.17.0.3
	/zammaddockercompose_zammad-memcached_1 172.17.0.2
	```

	As you can see, `zammaddockercompose_zammad-init_1` initially has the IP `172.17.0.5` (while it is running), but then loses it (when it exits).
	
4. Restart the WebSocket container:

	```
	zammad.ourcompany.tld ~/zammad-docker-compose # docker-compose restart zammad-websocket
	Restarting zammaddockercompose_zammad-websocket_1 ... done
	```
	
5. Now the WebSocket container has the IP `172.17.0.5`, and this in turn breaks NGINX:

	```
	zammad.ourcompany.tld ~ # dip
	/zammaddockercompose_zammad-nginx_1 172.17.0.10
	/zammaddockercompose_zammad-scheduler_1 172.17.0.9
	/zammaddockercompose_zammad-websocket_1 172.17.0.5
	/zammaddockercompose_zammad-backup_1 172.17.0.7
	/zammaddockercompose_zammad-railsserver_1 172.17.0.6
	/zammaddockercompose_zammad-init_1
	/zammaddockercompose_zammad-elasticsearch_1 172.17.0.4
	/zammaddockercompose_zammad-postgresql_1 172.17.0.3
	/zammaddockercompose_zammad-memcached_1 172.17.0.2
	```

**Comment by monotek** [comment]
Why do you restart the websocket container manually?

**Comment by martinvonwittich** [comment]
> Why do you restart the websocket container manually?

I don't; it crashed (check Ticket#1037760 for details), and I noticed that it remained broken afterwards even though it shouldn't. The manual restart here is just to illustrate how to reproduce the issue.

**Comment by monotek** [comment]
Is there an Github issue available too?
I don't have access to the zammad.com ticket system.

**Comment by martinvonwittich** [comment]
> Is there an Github issue available too?

No, not yet. I've sent you a PM on community.zammad.com as GitHub doesn't have a way to send direct messages :/

**Comment by monotek** [comment]
Sorry, but i'm not interested in discussing this privatly as it will not help others.

**Comment by martinvonwittich** [comment]
> Sorry, but i'm not interested in discussing this privatly as it will not help others.

I'm really sorry for making this so annoying, but it's a security issue that I don't want to make public before it's fixed :(

**Comment by monotek** [comment]
No, let it open for reference so we can link the other issue later and check if the problem is solved :-)

**Comment by martinvonwittich** [comment]
I think we just encountered another variation of this issue - our websocket container is broken. The symptoms were: a) [suddenly Zammad was no longer complaining about multiple tabs](https://community.zammad.org/t/multitab-functionality/1469/5?u=martin.von.wittich) (very much to the delight of our agents ^^) and b) Zammad was no longer refreshing anything (not so very much to the delight of anyone). Apparently postgres has restarted and probably received a new IP, and now the websocket container is attempting to connect to the old IP:

```
zammad ~ # docker ps
CONTAINER ID        IMAGE                                                       COMMAND                  CREATED             STATUS              PORTS                  NAMES
2bf812c7832f        zammad/zammad-docker-compose:zammad-postgresql-3.0.0-9      "/usr/local/bin/back…"   5 days ago          Up 5 days           5432/tcp               zammaddockercompose_zammad-backup_1
aac3849928c3        zammad/zammad-docker-compose:zammad-3.0.0-9                 "/docker-entrypoint.…"   5 days ago          Up 5 days           0.0.0.0:8080->80/tcp   zammaddockercompose_zammad-nginx_1
c1df4ff33168        zammad/zammad-docker-compose:zammad-3.0.0-9                 "/docker-entrypoint.…"   5 days ago          Up 5 days                                  zammaddockercompose_zammad-websocket_1
5518aa04d9c2        zammad/zammad-docker-compose:zammad-3.0.0-9                 "/docker-entrypoint.…"   5 days ago          Up 5 hours                                 zammaddockercompose_zammad-scheduler_1
dae88ac66b43        zammad/zammad-docker-compose:zammad-3.0.0-9                 "/docker-entrypoint.…"   5 days ago          Up 5 days                                  zammaddockercompose_zammad-railsserver_1
a17e86547ce2        zammad/zammad-docker-compose:zammad-postgresql-3.0.0-9      "docker-entrypoint.s…"   5 days ago          Up 24 hours         5432/tcp               zammaddockercompose_zammad-postgresql_1
19fa30d60c39        zammad/zammad-docker-compose:zammad-elasticsearch-3.0.0-9   "/bin/bash bin/es-do…"   5 days ago          Up 5 days           9200/tcp, 9300/tcp     zammaddockercompose_zammad-elasticsearch_1
278378b48f2d        zammad/zammad-docker-compose:zammad-memcached-3.0.0-9       "/docker-entrypoint.…"   5 days ago          Up 5 days           11211/tcp              zammaddockercompose_zammad-memcached_1
```

```
zammad ~ # docker logs --tail 1000 zammaddockercompose_zammad-websocket_1
[...]
E, [2019-06-26T07:48:42.569125 #1-47327053803360] ERROR -- : #<ActiveRecord::StatementInvalid: PG::UnableToSend: no connection to the server
: SELECT  "settings"."updated_at" FROM "settings" ORDER BY "settings"."updated_at" DESC LIMIT $1>
E, [2019-06-26T07:48:42.569239 #1-47327053803360] ERROR -- : ["/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:635:in `exec_prepared'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:635:in `block (2 levels) in exec_cache'", "/usr/local/bundle/gems/activesupport-5.1.7/lib/active_support/dependencies/interlock.rb:46:in `block in permit_concurrent_loads'", "/usr/local/bundle/gems/activesupport-5.1.7/lib/active_support/concurrency/share_lock.rb:185:in `yield_shares'", "/usr/local/bundle/gems/activesupport-5.1.7/lib/active_support/dependencies/interlock.rb:45:in `permit_concurrent_loads'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:634:in `block in exec_cache'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract_adapter.rb:613:in `block (2 levels) in log'", "/usr/local/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract_adapter.rb:612:in `block in log'", "/usr/local/bundle/gems/activesupport-5.1.7/lib/active_support/notifications/instrumenter.rb:21:in `instrument'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract_adapter.rb:604:in `log'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:633:in `exec_cache'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:613:in `execute_and_clear'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql/database_statements.rb:79:in `exec_query'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/database_statements.rb:375:in `select_prepared'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/database_statements.rb:40:in `select_all'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/query_cache.rb:97:in `select_all'", "/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/relation/calculations.rb:176:in `pluck'", "/opt/zammad/app/models/setting.rb:104:in `load'", "/opt/zammad/app/models/setting.rb:55:in `get'", "/opt/zammad/lib/app_version.rb:18:in `get'", "/opt/zammad/lib/app_version.rb:76:in `event_data'", "/opt/zammad/lib/sessions/event/login.rb:19:in `run'", "/opt/zammad/lib/sessions/event.rb:15:in `run'", "/opt/zammad/lib/websocket_server.rb:102:in `onmessage'", "/opt/zammad/lib/websocket_server.rb:25:in `block (3 levels) in run'", "/usr/local/bundle/gems/em-websocket-0.5.1/lib/em-websocket/connection.rb:18:in `trigger_on_message'", "/usr/local/bundle/gems/em-websocket-0.5.1/lib/em-websocket/message_processor_06.rb:52:in `message'", "/usr/local/bundle/gems/em-websocket-0.5.1/lib/em-websocket/framing07.rb:118:in `process_data'", "/usr/local/bundle/gems/em-websocket-0.5.1/lib/em-websocket/handler.rb:47:in `receive_data'", "/usr/local/bundle/gems/em-websocket-0.5.1/lib/em-websocket/connection.rb:76:in `receive_data'", "/usr/local/bundle/gems/eventmachine-1.2.7/lib/eventmachine.rb:195:in `run_machine'", "/usr/local/bundle/gems/eventmachine-1.2.7/lib/eventmachine.rb:195:in `run'", "/opt/zammad/lib/websocket_server.rb:10:in `run'", "script/websocket-server.rb:130:in `<top (required)>'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/cli/exec.rb:74:in `load'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/cli/exec.rb:74:in `kernel_load'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/cli/exec.rb:28:in `run'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/cli.rb:424:in `exec'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/cli.rb:27:in `dispatch'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/cli.rb:18:in `start'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/exe/bundle:30:in `block in <top (required)>'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'", "/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.6/exe/bundle:22:in `<top (required)>'", "/usr/local/bin/bundle:23:in `load'", "/usr/local/bin/bundle:23:in `<main>'"]
[...]
```

```
zammad ~ # docker logs --tail 1000 zammaddockercompose_zammad-websocket_1 | grep ERROR | wc -l   
26
```

**Comment by monotek** [comment]
How was the postgres container restartet?

**Comment by martinvonwittich** [comment]
> How was the postgres container restartet?

Manually (sorry, should have mentioned that, but I actually had forgotten today that I had restarted it yesterday...):

```
2019-06-25 10:17:33 martin.von.wittich    <zsh > docker restart zammaddockercompose_zammad-postgresql_1
```

The cause for this restart was that I had set `WEB_CONCURRENCY=4` for the railsserver on 2019-06-20:

```
zammad ~ # cat /root/zammad-docker-compose/docker-compose.override.yml
version: '2'

services:
  zammad-nginx:
    ports:
      - "8080:80"

  zammad-backup:
    environment:
      - NO_FILE_BACKUP=yes

  # https://community.zammad.org/t/memory-usage-in-zammad/373/2
  zammad-scheduler:
    environment:
      - RUBY_GC_MALLOC_LIMIT=1077216
      - RUBY_GC_MALLOC_LIMIT_MAX=2177216
      - RUBY_GC_OLDMALLOC_LIMIT=3000100
      - RUBY_GC_OLDMALLOC_LIMIT_MAX=4000100

  zammad-railsserver:
    environment:
      - WEB_CONCURRENCY=4

networks:
  default:
    driver_opts:
      com.docker.network.bridge.name: br-zammad
    ipam:
      driver: default
      config:
        - subnet: 172.17.0.0/24
```

That in turn had caused "sorry, too many clients already" errors in postgres because the default of `max_connections = 100` was obviously not enough for up to 4*30 = 120 puma threads. I edited `/var/lib/docker/volumes/zammaddockercompose_postgresql-data/_data/postgresql.conf` to increase `max_connections` to 200, and then I restarted postgres to apply the change.

So yeah, self-inflicted, but IMO still something that should work :)

**Comment by monotek** [comment]
so you think the zammad app should monitor the docker containers for restarts?
imho this is far out of scope...

**Comment by martinvonwittich** [comment]
> so you think the zammad app should monitor the docker containers for restarts and trigger restarts of the other containers by itself?

No, of course not. The app should ideally not even be aware that it's running inside a Docker container.

What I think should happen is:

- the app knows that the database is located on the host `zammad-postgresql` (because `config/database.yml` says so).
- when the connection to the database breaks, the app should reconnect to this host. But it should reconnect to the host `zammad-postgresql` (ie re-resolve the hostname to the current IP) and it should not attempt to reconnect to the old IP.
- as the DNS resolution is running through the Docker daemon, `zammad-postgresql` should correctly resolve to the current IP of that container.

Honestly I'm not really sure why it didn't work in the first place, but I also don't really know how Rails connects to the database. It does seem like it's caching the postgres IP somewhere and reuses this when reconnecting, or else the issue should never have occurred? I'll try to investigate.

**Comment by monotek** [comment]
As its in a linux container the resolved ip is not cached and i can't imagine rails would do that by itself.
Imho this is a docker / docker compose internal problem which can't be solved by the app.

**Comment by jeffmalterre** [comment]
I had everything setup perfectly using instructions from @t3easy in another thread to run zammad behind traefik using docker-frontend.

My machine rebooted and now I'm getting a 404 - is it normal for a container to have 2 ips?

/zammad-docker-compose_zammad-nginx_1 172.24.0.3172.27.0.10
/zammad-docker-compose_zammad-scheduler_1 172.27.0.9
/zammad-docker-compose_zammad-backup_1 172.27.0.8
/zammad-docker-compose_zammad-websocket_1 172.27.0.7
/zammad-docker-compose_zammad-railsserver_1 172.27.0.6
/zammad-docker-compose_zammad-init_1 172.27.0.5
/zammad-docker-compose_zammad-postgresql_1 172.27.0.4
/zammad-docker-compose_zammad-elasticsearch_1 172.27.0.3
/zammad-docker-compose_zammad-memcached_1 172.27.0.2
/frontend 172.24.0.2

**Comment by martinvonwittich** [comment]
> I'm really sorry for making this so annoying, but it's a security issue that I don't want to make public before it's fixed :(

For the record: the vulnerability was [ZAA-2020-06](https://zammad.com/news/security-advisory-zaa-2020-06) which allowed attackers to crash the websocket server by sending e.g. a literal `0`.
